parallel_algorithms/P3179/P3179.md

```diff
@@ -0,0 +1,6630 @@
+  constexpr bool ranges::ends_with(R1&& r1, R2&& r2, Pred pred = {},
+                                   Proj1 proj1 = {}, Proj2 proj2 = {});
+```
+Let `N1` be `ranges​::​distance(r1)` and `N2` be `ranges​::​distance(r2)`.
+
+*Returns*: `false` if `N1 < N2`, otherwise `ranges::equal(views::drop(ranges::ref_view(r1), N1 - static_cast<decltype(N1)>(N2)), r2, pred, proj1, proj2)`
+
+::: add
+```cpp
+template<@_execution-policy_@ Ep, @_random-access-sized-range_@ R1, @_random-access-sized-range_@ R2,
+         class Pred = ranges::equal_to, class Proj1 = identity, class Proj2 = identity>
+  requires indirectly_comparable<iterator_t<R1>, iterator_t<R2>, Pred, Proj1, Proj2>
+  bool ranges::ends_with(Ep&& exec, R1&& r1, R2&& r2,
+                         Pred pred = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
+```
+
+Let `N1` be `ranges​::​distance(r1)` and `N2` be `ranges​::​distance(r2)`.
+
+*Returns*: `false` if `N1 < N2`, otherwise `ranges::equal(std::forward<Ep>, views::drop(ranges::ref_view(r1), N1 - static_cast<decltype(N1)>(N2)), r2, pred, proj1, proj2)`
+:::
+
+## Modify `copy` in [alg.copy]{- .sref} ## {#modify_copy}
+
+```cpp
+template<class ExecutionPolicy, class ForwardIterator1, class ForwardIterator2>
+  ForwardIterator2 copy(ExecutionPolicy&& policy,
+                        ForwardIterator1 first, ForwardIterator1 last,
+                        ForwardIterator2 result);
+```
+
+::: add
+```cpp
+template<@_execution-policy_@ Ep, random_access_iterator I, sized_sentinel_for<I> S,
+         random_access_iterator O, sized_sentinel_for<O> OutS>
+  requires indirectly_copyable<I, O>
+  ranges::copy_result<I, O> ranges::copy(Ep&& exec, I first, S last, O result, OutS result_last);
+template<@_execution-policy_@ Ep, @_random-access-sized-range_@ R, @_random-access-sized-range_@ OutR>
+  requires indirectly_copyable<iterator_t<R>, iterator_t<OutR>>
+  ranges::copy_result<borrowed_iterator_t<R>, borrowed_iterator_t<OutR>>
+    ranges::copy(Ep&& exec, R&& r, OutR&& result);
+```
+:::
+
+[6]{.pnum} *Preconditions*: The ranges `[first, last)` and `[result, result + (last - first))` do not overlap.
+
+[7]{.pnum} *Effects*: Copies elements in the range `[first, last)` into the range `[result, result + (last - first))`.
+For each non-negative integer `n < (last - first)`, performs `*(result + n) = *(first + n)`.
+
+[8]{.pnum} *Returns*: `result + (last - first)`.
+
+[9]{.pnum} *Complexity*: Exactly `last - first` assignments.
+
+```cpp
+template<input_iterator I, weakly_incrementable O>
+  requires indirectly_copyable<I, O>
+  constexpr ranges::copy_n_result<I, O>
+    ranges::copy_n(I first, iter_difference_t<I> n, O result);
+```
+
+::: add
+```cpp
+template<@_execution-policy_@ Ep, random_access_iterator I, random_access_iterator O>
+  requires indirectly_copyable<I, O>
+  ranges::copy_n_result<I, O>
+    ranges::copy_n(Ep&& exec, I first, iter_difference_t<I> n, O result);
```

Comment:
We have discussed it with @rarutyun and have decided that it should not, on the ground that there is no input sentinel either, and also for other similar algorithms such as `fill_n` providing both a sentinel and a numerical limit would be redundant.

Reply:
But maybe we should consider the second numerical limit for the output.

Reply:
To be honest, I don't know. Really, I consider all the algorithms with _n (even without output as the separate parameter at all, like `fill_n`) unsafe.

For the given API:

```cpp
template< class O, class T = std::iter_value_t<O> >
requires std::output_iterator<O, const T&>
constexpr O fill_n( O first, std::iter_difference_t<O> n, const T& value );
```

`n` parameter says how much elements to fill but it doesn't say any  of the sequence.

The truly safe API would look like that (IMO):

```cpp
template< class O, sentinel_for<O> S, class T = std::iter_value_t<O> >
requires std::output_iterator<O, const T&>
constexpr O fill_n( O first, S last, std::iter_difference_t<O> n, const T& value );
```
where `S` is a sequence bound, and `n` has the same meaning but is `(last - first) < n` algorithm doesn't write out of bounds.

Originally, I thought that we should not touch algorithms with `_n` at all and keep them as close as possible to the serial range algorithms. However, we perhaps should add sentinel-for-output for at least `copy_n` and `move_n` to be consistent with `uninitialized_` counterparts. We already refer to them (`uninitialized_` counterparts ) in https://wg21.link/P3490 but the consistency is the only reason to add. I still believe that for `copy_n` input is unsafe without its own sentinel.

Honestly, I would teach the community to use `std::ranges::copy` with `take_view` instead of `copy_n`  (unless you convince me otherwise) if they want truly safe version of parallel algorithms.